include/libcyphal/transport/can/transport.hpp

```diff
@@ -87,11 +86,69 @@ class TransportImpl final : public ICanTransport
 
     void run(const TimePoint) override {}
 
+private:
+    // Until "canardMemFree must provide size" issue #216 is resolved,
+    // we need to store the size of the memory allocated.
+    // TODO: Remove this workaround when the issue is resolved.
+    // see https://github.com/OpenCyphal/libcanard/issues/216
+    //
+    struct CanardMemory
+    {
+        struct Layout
+        {
+            std::size_t size;
+            alignas(std::max_align_t) cetl::byte data[sizeof(std::max_align_t)];
+        };
+
+        static constexpr std::size_t DataOffset = sizeof(Layout) - sizeof(Layout::data);
+    };
+
+    CETL_NODISCARD static inline TransportImpl& getSelfFrom(const CanardInstance* const ins)
+    {
+        CETL_DEBUG_ASSERT(ins != nullptr, "Expected canard instance.");
+        CETL_DEBUG_ASSERT(ins->user_reference != nullptr, "Expected `this` transport as user reference.");
+
+        return *static_cast<TransportImpl*>(ins->user_reference);
+    }
+
+    CETL_NODISCARD static void* canardMemoryAllocate(CanardInstance* ins, size_t amount)
+    {
+        auto& self = getSelfFrom(ins);
+
+        const auto memory_size   = CanardMemory::DataOffset + amount;
+        const auto canard_memory = static_cast<CanardMemory::Layout*>(self.memory_.allocate(memory_size));
+        if (canard_memory == nullptr)
+        {
+            return nullptr;
+        }
+        canard_memory->size = memory_size;
+
+        return &canard_memory->data;
+    }
+
+    static void canardMemoryFree(CanardInstance* ins, void* pointer)
+    {
+        if (pointer == nullptr)
+        {
+            return;
+        }
+
+        const auto uint_ptr = reinterpret_cast<std::uintptr_t>(pointer);
+        CETL_DEBUG_ASSERT(uint_ptr > CanardMemory::DataOffset, "Invalid too small pointer.");
+        const auto canard_memory = reinterpret_cast<CanardMemory::Layout*>(uint_ptr - CanardMemory::DataOffset);
```

Comment:
The conversion to integer and the DataOffset all look a bit shady but we will accept this for now pending proper solution. I think it would have been a lot easier to just go like:

```c++
struct CanardMemory
{
    alignas(std::max_align_t) std::size_t size;
};
static_assert(sizeof(CanardMemory) == alignof(std::max_align_t));
```

And then:

```c++
auto* const mem = static_cast<CanardMemory*>(pointer)-1;
```

But it's good enough as a tentative solution.